### ISSUE TITLE
[Feature Request] More general metrics

### ISSUE THREAD

**Issue by AntoinePrv** [pytorch/ignite#218]
I find the metrics to be a bit limited as one might want to pass additional options (even tensors) to the loss.
For instance in recurrent models with different sequence lengths, one would use a mask to avoid counting errors on padded time steps.
The mask is necessary in the loss to know which outputs to use in the final averaging/ loss.

**Comment by colllin** [comment]
Could you maybe give an example using code or pseudo-code?

**Comment by vfdev-5** [comment]
@AntoinePrv could you please give an example on this issue ? 

If this becomes stale, maybe we can close this...

**Comment by AntoinePrv** [comment]
Sorry, I've postponed answering this for far too long...

Ideally we would want `ignite.metric.Loss` to be able to wrap any loss function from PyTorch, right?
Many loss functions have a `weights` attribute (tensor) to scale the different element (over the batch for instance).

Although the argument can be set statically before passing `loss_fn` to the `Loss`, this cannot be used dynamically (`weights` to depend on batch) with the current class (at least not nicely).
This is because of [this line](https://github.com/pytorch/ignite/blob/c71a3b04d2514f27b401d11a848115bebbae19d8/ignite/metrics/loss.py#L23): eventually only two arguments are extracted out of the output.

Something like the following might be an idea:
```python
args, kwargs = output
average_loss = self._loss_fn(*args, **kwargs)
```

(Note: this is also the case with the masks I was mentioning in the first comment, but we can stick to the example of `weights` as the point is the same)

**Comment by vfdev-5** [comment]
@AntoinePrv coming back on what you say about `weights`. Following the [cross entropy docs](https://pytorch.org/docs/stable/nn.html#torch.nn.functional.cross_entropy), `weights`  rescales classes and not samples... 
If I understand correctly you would like to be able to apply weights coming from batch generator ?

**Comment by vfdev-5** [comment]
@AntoinePrv we discussed a similar point [here](https://github.com/pytorch/ignite/pull/112#issuecomment-373838309). Yes, I would say your remark totally make sense, just for instance, we have findout if we could cover this with our vanilla `Loss` or create somelthing like `LossWithArgs` ... Let me think about it.

**Comment by AntoinePrv** [comment]
You would change the `output_transform` default to `lambda x: (x, {})` so that the evaluator still returns `y_pred, y`.

Note that using that loss would require rewriting a custom evaluator to change the output.

**Comment by vfdev-5** [comment]
@AntoinePrv a hacky solution can be something similar:
```python
import torch.nn.functional as F

def loss_fn(y_pred, target):
    y, kwargs = target    
    return F.cross_entropy(y_pred, y, **kwargs)

metrics = {
    'loss': Loss(loss_fn, output_transform=lambda y_pred, y, kwargs: (y_pred, (y, kwargs)))
}
```

**Comment by AntoinePrv** [comment]
Sure, I also made my own solution. I just wanted to know if you're interested in making this change.

**Comment by AntoinePrv** [comment]
Partially, however, it's hard to say because all of the tests involve:
```python
loss.update((y_pred, y))
```
I could succeed the tests by replacing it with
```python
loss.update(loss.update(((y_pred, y), {})))
```

I rely on changing the default argument to the constructor from `output_transform=lambda x: x` to `output_transform=lambda x: (x, {})`, which means it will not be backward compatible for any user having used that input.

__Note__:
I was too greedy with `args, kwargs = output` because `y` is used a few line after for:
```python
self._sum += average_loss.item() * y.shape[0]
self._num_examples += y.shape[0]
```
A sufficient change could be `(y_pred, y), kwargs = output`

**Comment by jasonkriss** [comment]
I agree that this is a completely valid point and something we should address. If we want to be backwards compatible, I think the easiest thing to do is to change [this](https://github.com/pytorch/ignite/blob/master/ignite/metrics/loss.py#L23) to something roughly like this:

```python
y_pred, y = output[0], output[1]
kwargs = output[2] if len(output) > 2 else {}
```

**Comment by alykhantejani** [comment]
lets get this into 0.1.1
